File: Vanta Speech/Vanta Speech/Core/Calendar/API/GraphCalendarService.swift
import Foundation

/// Сервис для работы с календарём через Microsoft Graph API
final class GraphCalendarService {

    private let baseURL = "https://graph.microsoft.com/v1.0"
    private let authManager: MSALAuthManager

    init(authManager: MSALAuthManager) {
        self.authManager = authManager
    }

    // MARK: - Private Helpers

    /// Создание авторизованного запроса
    private func authorizedRequest(url: URL, method: String = "GET") async throws -> URLRequest {
        let token = try await authManager.acquireTokenSilently()

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        return request
    }

    /// Выполнение запроса с декодированием ответа
    private func performRequest<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw GraphError.invalidResponse
        }

        // Обработка ошибок HTTP
        switch httpResponse.statusCode {
        case 200...299:
            break
        case 401:
            throw GraphError.unauthorized
        case 403:
            throw GraphError.forbidden
        case 404:
            throw GraphError.notFound
        case 429:
            // Rate limiting
            let retryAfter = httpResponse.value(forHTTPHeaderField: "Retry-After")
            throw GraphError.rateLimited(retryAfter: Int(retryAfter ?? "60") ?? 60)
        default:
            throw GraphError.httpError(statusCode: httpResponse.statusCode, data: data)
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        return try decoder.decode(T.self, from: data)
    }
}

// MARK: - Fetch Events

extension GraphCalendarService {

    /// Получить события календаря за период
    /// - Parameters:
    ///   - startDate: Начало периода
    ///   - endDate: Конец периода
    ///   - maxResults: Максимальное количество событий
    /// - Returns: Массив событий
    func fetchEvents(
        from startDate: Date,
        to endDate: Date,
        maxResults: Int = 50
    ) async throws -> [GraphEvent] {

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]

        let startString = formatter.string(from: startDate)
        let endString = formatter.string(from: endDate)

        // calendarView раскрывает recurring events в отдельные instances
        var components = URLComponents(string: "\(baseURL)/me/calendarView")!
        components.queryItems = [
            URLQueryItem(name: "startDateTime", value: startString),
            URLQueryItem(name: "endDateTime", value: endString),
            URLQueryItem(name: "$select", value: "id,subject,start,end,attendees,organizer,isOrganizer,iCalUId,bodyPreview,webLink,location"),
            URLQueryItem(name: "$orderby", value: "start/dateTime"),
            URLQueryItem(name: "$top", value: String(maxResults))
        ]

        guard let url = components.url else {
            throw GraphError.invalidResponse
        }

        let request = try await authorizedRequest(url: url)
        let response: GraphEventsResponse = try await performRequest(request)

        return response.value
    }

    /// Получить события на сегодня
    func fetchTodayEvents() async throws -> [GraphEvent] {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: Date())
        let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!

        return try await fetchEvents(from: startOfDay, to: endOfDay)
    }

    /// Получить конкретное событие по ID
    func fetchEvent(id: String) async throws -> GraphEvent {
        let url = URL(string: "\(baseURL)/me/events/\(id)")!
        let request = try await authorizedRequest(url: url)
        return try await performRequest(request)
    }
}

// MARK: - Create Event

extension GraphCalendarService {

    /// Создать новое событие
    /// - Parameter event: Данные события
    /// - Returns: Созданное событие
    func createEvent(_ event: CreateEventRequest) async throws -> GraphEvent {
        let url = URL(string: "\(baseURL)/me/events")!
        var request = try await authorizedRequest(url: url, method: "POST")

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        request.httpBody = try encoder.encode(event)

        return try await performRequest(request)
    }

    /// Быстрое создание события
    /// - Parameters:
    ///   - title: Название события
    ///   - start: Время начала
    ///   - end: Время окончания
    ///   - location: Место проведения (опционально)
    /// - Returns: Созданное событие
    func createEvent(
        title: String,
        start: Date,
        end: Date,
        location: String? = nil
    ) async throws -> GraphEvent {
        let request = CreateEventRequest(
            subject: title,
            start: start,
            end: end,
            location: location
        )
        return try await createEvent(request)
    }
}

// MARK: - Update Event

extension GraphCalendarService {

    /// Обновить существующее событие (частичное обновление)
    /// - Parameters:
    ///   - id: ID события
    ///   - changes: Изменения
    /// - Returns: Обновлённое событие
    func updateEvent(id: String, changes: UpdateEventRequest) async throws -> GraphEvent {
        let url = URL(string: "\(baseURL)/me/events/\(id)")!
        var request = try await authorizedRequest(url: url, method: "PATCH")

        let encoder = JSONEncoder()
        request.httpBody = try encoder.encode(changes)

        return try await performRequest(request)
    }

    /// Переименовать событие
    /// - Parameters:
    ///   - id: ID события
    ///   - newTitle: Новое название
    /// - Returns: Обновлённое событие
    func renameEvent(id: String, newTitle: String) async throws -> GraphEvent {
        var changes = UpdateEventRequest()
        changes.subject = newTitle
        return try await updateEvent(id: id, changes: changes)
    }
}

// MARK: - Delete Event

extension GraphCalendarService {

    /// Удалить событие
    /// - Parameter id: ID события
    func deleteEvent(id: String) async throws {
        let url = URL(string: "\(baseURL)/me/events/\(id)")!
        let request = try await authorizedRequest(url: url, method: "DELETE")

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 204 else {
            throw GraphError.deleteFailed
        }
    }
}

// MARK: - Attendees

extension GraphCalendarService {

    /// Получить участников события с их статусами
    /// - Parameter eventId: ID события
    /// - Returns: Список участников
    func getAttendees(eventId: String) async throws -> [MeetingParticipant] {
        let event = try await fetchEvent(id: eventId)

        var participants: [MeetingParticipant] = []

        // Организатор
        if let organizer = event.organizer {
            participants.append(MeetingParticipant(
                email: organizer.emailAddress.address,
                name: organizer.emailAddress.name,
                role: .organizer,
                responseStatus: .accepted  // Организатор всегда accepted
            ))
        }

        // Участники
        for attendee in event.attendees ?? [] {
            let role: ParticipantRole = attendee.type == "required" ? .required : .optional
            let status = mapResponseStatus(attendee.status?.response)

            participants.append(MeetingParticipant(
                email: attendee.emailAddress.address,
                name: attendee.emailAddress.name,
                role: role,
                responseStatus: status
            ))
        }

        return participants
    }

    private func mapResponseStatus(_ response: String?) -> ParticipantResponseStatus {
        switch response {
        case "accepted": return .accepted
        case "declined": return .declined
        case "tentativelyAccepted": return .tentative
        case "none", .none: return .notResponded
        default: return .notResponded
        }
    }
}

// MARK: - User Profile

extension GraphCalendarService {

    /// Получить профиль текущего пользователя
    func fetchUserProfile() async throws -> UserProfile {
        let url = URL(string: "\(baseURL)/me")!
        let request = try await authorizedRequest(url: url)
        return try await performRequest(request)
    }
}

/// Профиль пользователя Microsoft
struct UserProfile: Codable {
    let id: String
    let displayName: String?
    let mail: String?
    let userPrincipalName: String?

    var email: String {
        mail ?? userPrincipalName ?? "unknown"
    }
}

File: Vanta Sppech Android/app/src/main/java/com/vanta/speech/core/calendar/GraphCalendarService.kt
package com.vanta.speech.core.calendar

import android.util.Log
import com.vanta.speech.core.calendar.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for working with Microsoft Graph Calendar API
 * Android equivalent of iOS GraphCalendarService
 */
@Singleton
class GraphCalendarService @Inject constructor(
    private val authManager: MSALAuthManager
) {
    companion object {
        private const val TAG = "GraphCalendarService"
        private const val BASE_URL = "https://graph.microsoft.com/v1.0"
    }

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        isLenient = true
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    /**
     * Fetch calendar events for a date range
     */
    suspend fun fetchEvents(
        from: Date,
        to: Date,
        maxResults: Int = 50
    ): List<GraphEvent> = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")

        val startString = formatter.format(from)
        val endString = formatter.format(to)

        val url = buildString {
            append("$BASE_URL/me/calendarView")
            append("?startDateTime=$startString")
            append("&endDateTime=$endString")
            append("&\$select=id,subject,start,end,attendees,organizer,isOrganizer,iCalUId,bodyPreview,webLink,location")
            append("&\$orderby=start/dateTime")
            append("&\$top=$maxResults")
        }

        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            handleHttpError(response.code, response.body?.string())
        }

        val responseBody = response.body?.string() ?: throw GraphError.InvalidResponse
        val eventsResponse = json.decodeFromString<GraphEventsResponse>(responseBody)

        Log.d(TAG, "Fetched ${eventsResponse.value.size} events")
        eventsResponse.value
    }

    /**
     * Fetch today's events
     */
    suspend fun fetchTodayEvents(): List<GraphEvent> {
        val calendar = Calendar.getInstance()
        val startOfDay = calendar.apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.time

        val endOfDay = calendar.apply {
            add(Calendar.DAY_OF_MONTH, 1)
        }.time

        return fetchEvents(startOfDay, endOfDay)
    }

    /**
     * Fetch a specific event by ID
     */
    suspend fun fetchEvent(id: String): GraphEvent = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val request = Request.Builder()
            .url("$BASE_URL/me/events/$id")
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            handleHttpError(response.code, response.body?.string())
        }

        val responseBody = response.body?.string() ?: throw GraphError.InvalidResponse
        json.decodeFromString<GraphEvent>(responseBody)
    }

    /**
     * Create a new event
     */
    suspend fun createEvent(event: CreateEventRequest): GraphEvent = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val eventJson = json.encodeToString(event)

        val request = Request.Builder()
            .url("$BASE_URL/me/events")
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .post(eventJson.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            handleHttpError(response.code, response.body?.string())
        }

        val responseBody = response.body?.string() ?: throw GraphError.InvalidResponse
        json.decodeFromString<GraphEvent>(responseBody)
    }

    /**
     * Update an existing event
     */
    suspend fun updateEvent(id: String, changes: UpdateEventRequest): GraphEvent = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val changesJson = json.encodeToString(changes)

        val request = Request.Builder()
            .url("$BASE_URL/me/events/$id")
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .patch(changesJson.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            handleHttpError(response.code, response.body?.string())
        }

        val responseBody = response.body?.string() ?: throw GraphError.InvalidResponse
        json.decodeFromString<GraphEvent>(responseBody)
    }

    /**
     * Delete an event
     */
    suspend fun deleteEvent(id: String): Unit = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val request = Request.Builder()
            .url("$BASE_URL/me/events/$id")
            .addHeader("Authorization", "Bearer $token")
            .delete()
            .build()

        val response = client.newCall(request).execute()

        if (response.code != 204) {
            handleHttpError(response.code, response.body?.string())
        }
    }

    /**
     * Fetch user profile
     */
    suspend fun fetchUserProfile(): UserProfile = withContext(Dispatchers.IO) {
        val token = authManager.acquireTokenSilently()

        val request = Request.Builder()
            .url("$BASE_URL/me")
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            handleHttpError(response.code, response.body?.string())
        }

        val responseBody = response.body?.string() ?: throw GraphError.InvalidResponse
        json.decodeFromString<UserProfile>(responseBody)
    }

    /**
     * Get attendees for an event
     */
    suspend fun getAttendees(eventId: String): List<MeetingParticipant> {
        val event = fetchEvent(eventId)

        val participants = mutableListOf<MeetingParticipant>()

        // Add organizer
        event.organizer?.let { organizer ->
            participants.add(
                MeetingParticipant(
                    email = organizer.emailAddress.address,
                    name = organizer.emailAddress.name,
                    role = ParticipantRole.ORGANIZER,
                    responseStatus = ParticipantResponseStatus.ACCEPTED
                )
            )
        }

        // Add attendees
        event.attendees?.forEach { attendee ->
            val role = when (attendee.type) {
                "required" -> ParticipantRole.REQUIRED
                else -> ParticipantRole.OPTIONAL
            }

            val status = when (attendee.status?.response) {
                "accepted" -> ParticipantResponseStatus.ACCEPTED
                "declined" -> ParticipantResponseStatus.DECLINED
                "tentativelyAccepted" -> ParticipantResponseStatus.TENTATIVE
                else -> ParticipantResponseStatus.NOT_RESPONDED
            }

            participants.add(
                MeetingParticipant(
                    email = attendee.emailAddress.address,
                    name = attendee.emailAddress.name,
                    role = role,
                    responseStatus = status
                )
            )
        }

        return participants
    }

    private fun handleHttpError(statusCode: Int, body: String?): Nothing {
        Log.e(TAG, "HTTP Error $statusCode: $body")
        throw when (statusCode) {
            401 -> GraphError.Unauthorized
            403 -> GraphError.Forbidden
            404 -> GraphError.NotFound
            429 -> GraphError.RateLimited(60)
            else -> GraphError.HttpError(statusCode, body)
        }
    }
}

/**
 * Graph API Errors
 */
sealed class GraphError : Exception() {
    data object InvalidResponse : GraphError() {
        private fun readResolve(): Any = InvalidResponse
        override val message = "Неверный ответ сервера"
    }

    data object Unauthorized : GraphError() {
        private fun readResolve(): Any = Unauthorized
        override val message = "Не авторизован"
    }

    data object Forbidden : GraphError() {
        private fun readResolve(): Any = Forbidden
        override val message = "Доступ запрещён"
    }

    data object NotFound : GraphError() {
        private fun readResolve(): Any = NotFound
        override val message = "Не найдено"
    }

    data object DeleteFailed : GraphError() {
        private fun readResolve(): Any = DeleteFailed
        override val message = "Ошибка удаления"
    }

    data class RateLimited(val retryAfterSeconds: Int) : GraphError() {
        override val message = "Превышен лимит запросов. Повторите через $retryAfterSeconds сек."
    }

    data class HttpError(val statusCode: Int, val body: String?) : GraphError() {
        override val message = "HTTP ошибка $statusCode: $body"
    }
}
